# Homework 1

## FINM 25000 - 2025

### UChicago Financial Mathematics

* Mark Hendricks
* hendricks@uchicago.edu

## HBS Case

### *The Harvard Management Company and Inflation-Indexed Bonds*

### Notation
(Hidden LaTeX commands)

$$\newcommand{\mux}{\tilde{\boldsymbol{\mu}}}$$
$$\newcommand{\wtan}{\boldsymbol{\text{w}}^{\text{tan}}}$$
$$\newcommand{\wtarg}{\boldsymbol{\text{w}}^{\text{port}}}$$
$$\newcommand{\mutarg}{\tilde{\boldsymbol{\mu}}^{\text{port}}}$$
$$\newcommand{\wEW}{\boldsymbol{\text{w}}^{\text{EW}}}$$
$$\newcommand{\wRP}{\boldsymbol{\text{w}}^{\text{RP}}}$$
$$\newcommand{\wREG}{\boldsymbol{\text{w}}^{\text{REG}}}$$

***

# 1. HMC's Approach

**Section 1 is not graded**, and you do not need to submit your answers. But you are encouraged to think about them, and we will discuss them.

### 1.
There are thousands of individual risky assets in which HMC can invest.  Explain why MV optimization across 1,000 securities is infeasible.

### 2.
Rather than optimize across all securities directly, HMC runs a two-stage optimization.
1. They build asset class portfolios with each one optimized over the securities of the specific asset class.  
2. HMC combines the asset-class portfolios into one total optimized portfolio.

In order for the two-stage optimization to be a good approximation of the full MV-optimization on all assets, what must be true of the partition of securities into asset classes?

### 3.
Should TIPS form a new asset class or be grouped into one of the other 11 classes?

### 4.
Why does HMC focus on real returns when analyzing its portfolio allocation? Is this just a matter of scaling, or does using real returns versus nominal returns potentially change the MV solution?

### 5.
The case discusses the fact that Harvard places bounds on the portfolio allocation rather than implementing whatever numbers come out of the MV optimization problem.

How might we adjust the stated optimization problem in the lecture notes to reflect the extra constraints Harvard is using in their bounded solutions given in Exhibits 5 and 6?

### 6.
Exhibits 5 shows zero allocation to domestic equities and domestic bonds across the entire computed range of targeted returns, (5.75% to 7.25%). Conceptually, why is the constraint binding in all these cases? What would the unconstrained portfolio want to do with those allocations and why?

### 7.
Exhibit 6 changes the constraints, (tightening them in most cases.) How much deterioration do we see in the mean-variance tradeoff that Harvard achieved?

***

# 2 Mean-Variance Optimization

<i>This section is graded for a good-faith effort by your group. Submit your write-up- along with your supporting code. </i>

### Data
You will need the file in the github repo, `data/multi_asset_etf_data.xlsx`.
- The time-series data gives monthly returns for the 11 asset classes and a short-term Treasury-bill fund return, ("SHV",) which we consider as the risk-free rate.
- The data is provided in total returns, (in which case you should ignore the SHV column,) as well as excess returns, (where SHV has been subtracted from the other columns.)
- These are nominal returns-they are not adjusted for inflation, and in our calculations we are not making any adjustment for inflation.
- The exhibit data that comes via Harvard with the case is unnecessary for our analysis.

### Model
We are going to analyze the problem in terms of **excess** returns.
- Thus, you will focus on the `Excess Returns` section of the lecture notes, especially the formulas on slide 50.
- Be sure to use the`excess returns` tab of the data.

### Format
In the questions below, **annualize the statistics** you report.
- Annualize the mean of monthly returns with a scaling of 12.
- Annualize the volatility of monthly returns with a scaling of $\sqrt{12}$
- The Sharpe Ratio is the mean return divided by the volatility of returns. Accordingly, we can annualize the Sharpe Ratio with a scaling of $\sqrt{12}$
- Note that we are not scaling the raw timeseries data, just the statistics computed from it (mean, vol, Sharpe).

### Footnotes

#### Data File
* The case does not give time-series data, so this data has been compiled outside of the case, and it intends to represent the main asset classes under consideration via various ETFs. For details on the specific securities/indexes, check the “Info” tab of the data.

#### Risk-free rate
* In the lecture-note we considered a constant risk-free rate. It is okay that our risk-free rate changes over time, but the assumption is that investors know it’s value one-period ahead of time. Thus, at any given point in time, it is a risk-free rate for the next period. (This is often discussed as the "bank account" or "money market account" in other settings.

### 1. Summary Statistics
* Calculate and display the mean and volatility of each asset’s excess return. (Recall we use volatility to refer to standard deviation.)
* Which assets have the best and worst Sharpe ratios? Recall that the Sharpe Ratio is simply the ratio of the mean-to-volatility of excess returns:
$$\text{sharpe ratio of investment }i = \frac{\mux_i}{\sigma_i}$$

Be sure to annualize all three statss (mean, vol, Sharpe).
* mean is scaled by `12`
* vol is scaled by `sqrt(12)`
* Sharpe is scaled by `sqrt(12)`


### 2. Descriptive Analysis
* Calculate the correlation matrix of the returns. Which pair has the highest correlation? And the lowest?
* How well have TIPS done in our sample? Have they outperformed domestic bonds? Foreign bonds?

### 3. The MV frontier.
* Compute and display the weights of the tangency portfolios: $\wtan$.
* Does the ranking of weights align with the ranking of Sharpe ratios?
* Compute the mean, volatility, and Sharpe ratio for the tangency portfolio corresponding to
$\wtan$.

### 4. TIPS
Assess how much the tangency portfolio (and performance) change if...
* TIPS are dropped completely from the investment set.
* The expected excess return to TIPS is adjusted to be 0.0012 higher than what the historic sample shows.

Based on the analysis, do TIPS seem to expand the investment opportunity set, implying that Harvard should consider them as a separate asset?

#Responses - Georgia Martin

In [2]:
from google.colab import drive
drive.mount('/content/drive')
import pandas as pd
hw1_data = pd.read_excel('/content/drive/My Drive/multi_asset_etf_data.xlsx', sheet_name='excess returns', index_col=0,parse_dates=[0])

Mounted at /content/drive


#1. Summary Statistics

In [3]:
import numpy as np
def performance_summary(df, adj_factor=12):
  """
  Generates performance summary containing mean, volatility (vol), and Sharpe ratio (sharpe)
  per each column in the given dataframe. Adj_factor is used to annualize the statistics.
  """
  summary = {}
  summary['mean'] = df.mean()*adj_factor #avg of each month * 12
  summary['vol'] = df.std()*np.sqrt(adj_factor)
  summary['sharpe'] = summary['mean']/summary['vol']
  return pd.DataFrame(summary)
metrics = performance_summary(hw1_data).sort_values('sharpe', ascending=False)
metrics.head(11)

mean       vol    sharpe
SPY  0.128141  0.142839  0.897103
HYG  0.041371  0.075928  0.544873
IYR  0.074916  0.168675  0.444143
PSP  0.092561  0.213370  0.433804
EFA  0.061775  0.150903  0.409372
TIP  0.020502  0.051115  0.401091
QAI  0.019327  0.049073  0.393838
IEF  0.016404  0.063442  0.258569
EEM  0.029339  0.176164  0.166542
DBC -0.005292  0.166553 -0.031774
BWX -0.007716  0.082789 -0.093202

In [4]:
print('Max sharpe ratio:', metrics['sharpe'].idxmax(), metrics['sharpe'].max())
print('Min sharpe ratio:', metrics['sharpe'].idxmin(), metrics['sharpe'].min())

Max sharpe ratio: SPY 0.897102626289479
Min sharpe ratio: BWX -0.09320174565672584


The maximum sharpe ratio belongs to SPY with a value of 0.9. The lowest belongs to BWX with a value of -0.09.

#2. Descriptive Analysis

In [5]:
corr_matrix = hw1_data.corr()
corr_matrix.head(11)

BWX       DBC       EEM       EFA       HYG       IEF       IYR  \
BWX  1.000000  0.191116  0.621673  0.602820  0.602555  0.580891  0.552557   
DBC  0.191116  1.000000  0.511667  0.500922  0.461887 -0.300207  0.280518   
EEM  0.621673  0.511667  1.000000  0.819925  0.691167  0.026704  0.584063   
EFA  0.602820  0.500922  0.819925  1.000000  0.787191  0.042639  0.699292   
HYG  0.602555  0.461887  0.691167  0.787191  1.000000  0.187258  0.739356   
IEF  0.580891 -0.300207  0.026704  0.042639  0.187258  1.000000  0.316532   
IYR  0.552557  0.280518  0.584063  0.699292  0.739356  0.316532  1.000000   
PSP  0.526692  0.453303  0.750109  0.895320  0.812157  0.022436  0.749836   
QAI  0.630276  0.475311  0.774697  0.847864  0.807893  0.179761  0.718529   
SPY  0.439994  0.432162  0.687751  0.845863  0.793518  0.000815  0.754711   
TIP  0.675151  0.109006  0.378792  0.394821  0.538648  0.754102  0.598742   

          PSP       QAI       SPY       TIP  
BWX  0.526692  0.630276  0.439994  0.675151  
DBC  0.453303  0.475311  0.432162  0.109006  
EEM  0.750109  0.774697  0.687751  0.378792  
EFA  0.895320  0.847864  0.845863  0.394821  
HYG  0.812157  0.807893  0.793518  0.538648  
IEF  0.022436  0.179761  0.000815  0.754102  
IYR  0.749836  0.718529  0.754711  0.598742  
PSP  1.000000  0.873395  0.891687  0.408005  
QAI  0.873395  1.000000  0.866845  0.516670  
SPY  0.891687  0.866845  1.000000  0.381625  
TIP  0.408005  0.516670  0.381625  1.000000

In [6]:
corr_unstacked = corr_matrix.where(np.triu(np.ones(corr_matrix.shape),
                                           k=1).astype(bool)).unstack().dropna()
highest_corr = corr_unstacked.idxmax(), corr_unstacked.max()
lowest_corr = corr_unstacked.idxmin(), corr_unstacked.min()
print("Highest correlation pair:", highest_corr)
print("Lowest correlation pair:", lowest_corr)

Highest correlation pair: (('PSP', 'EFA'), 0.8953201243752301)
Lowest correlation pair: (('IEF', 'DBC'), -0.30020714105644963)


EFA and PSP have the highest correlation of 0.89, and IEF and DBC have the lowest correlation of -0.3.

Comparing TIPS with IEF (Intermediate-Term U.S. Treasuries) and BWX (Foreign Government Bonds), we can see that TIPS delivered the strongest performance among the government bond ETFs analyzed. TIP had an annualized excess return of 2.05%, with a volatility of 5.11%, resulting in a Sharpe ratio of 0.40. In contrast, IEF had a lower return (1.64%) and a lower Sharpe ratio (0.26), indicating less attractive risk-adjusted performance. BWX, underperformed with a negative excess return of -0.77% and a negative Sharpe ratio (-0.09). This suggests that, in this sample, TIPS provided both higher returns and more efficient risk-reward tradeoffs than domestic and foreign government bonds.

***

# 3. The MV Fronteir

In [7]:
def tan_weights(df):
  """
  Given a df with excess returns, computes the weights of the tangency
  portfolios.
  """
  mu = df.mean().values  # mean excess returns (monthly)
  Sigma = df.cov().values  # covariance matrix (monthly)
  ones = np.ones(len(mu))

  inv_Sigma = np.linalg.inv(Sigma) #inverse

  top = inv_Sigma @ mu
  bottom = ones @ top
  w_tan = top / bottom
  tangency_weights = pd.Series(w_tan, index=df.columns)
  tangency_df = tangency_weights.to_frame(name="Tangency Portfolio Weights")
  return tangency_df

result = tan_weights(hw1_data)
result.sort_values(by='Tangency Portfolio Weights', ascending=False).head(11)

Tangency Portfolio Weights
SPY                   10.729782
IEF                    9.010591
HYG                    2.634922
TIP                    1.330595
EEM                    0.853591
EFA                    0.385285
DBC                   -0.112341
PSP                   -1.716842
IYR                   -2.382146
BWX                   -6.117430
QAI                  -13.616007

In [8]:
metrics.head(11)
merged = pd.merge(metrics, result, left_index=True, right_index=True)
merged.head(11)

mean       vol    sharpe  Tangency Portfolio Weights
SPY  0.128141  0.142839  0.897103                   10.729782
HYG  0.041371  0.075928  0.544873                    2.634922
IYR  0.074916  0.168675  0.444143                   -2.382146
PSP  0.092561  0.213370  0.433804                   -1.716842
EFA  0.061775  0.150903  0.409372                    0.385285
TIP  0.020502  0.051115  0.401091                    1.330595
QAI  0.019327  0.049073  0.393838                  -13.616007
IEF  0.016404  0.063442  0.258569                    9.010591
EEM  0.029339  0.176164  0.166542                    0.853591
DBC -0.005292  0.166553 -0.031774                   -0.112341
BWX -0.007716  0.082789 -0.093202                   -6.117430

While SPY remains the highest Sharpe ratio and tangency portfolio rate, the rest do not follow.

In [9]:
def tan_metrics(df_ereturns,df):
  """
  df_ereturns: original data
  df: tangency portfolio weights meged with metrics

  Given a dataframe of metrics including tangency weighted portfolios, computes
  the mean, volatility, and Sharpe ratio wrt the tangency weight.
  """
  w = df["Tangency Portfolio Weights"].values
  mu = df["mean"].values
  Sigma = df_ereturns.cov().values

  assert w.shape[0] == Sigma.shape[0], "Mismatch: weights and covariance matrix size"

  mean_portfolio = w @ mu
  vol_portfolio = np.sqrt(w.T @ Sigma @ w)
  sharpe_portfolio = mean_portfolio / vol_portfolio

  return pd.DataFrame({
      "Annualized Mean": [mean_portfolio],
      "Annualized Volatility": [vol_portfolio],
      "Annualized Sharpe Ratio": [sharpe_portfolio]
    })
tan_metrics(hw1_data, merged).head()

Annualized Mean  Annualized Volatility  Annualized Sharpe Ratio
0         1.155145               0.466146                 2.478073

For the tangency portfolio corresponding to
$\wtan$, the annualized mean is 1.16, the annualized volatility is 0.47, and the annualized sharpe ratio is 2.48.

#4. TIPS

Dropped TIPS:

In [10]:
dropped_tips = hw1_data.drop(columns=['TIP'])
metrics_no_tips = performance_summary(dropped_tips).sort_values('sharpe', ascending=False)
wtan_no_tip = tan_weights(dropped_tips)["Tangency Portfolio Weights"]
wtan_no_tip = pd.merge(metrics_no_tips, wtan_no_tip, left_index=True, right_index=True)
tan_metrics(dropped_tips, wtan_no_tip).head()

Annualized Mean  Annualized Volatility  Annualized Sharpe Ratio
0         1.570867               0.538566                 2.916757

All aspects (mean, vol, and sharpe) are higher without TIPS.

Adjusted TIPS:


In [11]:
adj_data = hw1_data.copy()
adj_data['TIP'] = adj_data['TIP'] + 0.0012
metrics_adj = performance_summary(adj_data).sort_values('sharpe', ascending=False)
wtan_adj = tan_weights(adj_data)["Tangency Portfolio Weights"]
wtan_adjusted = pd.merge(metrics_adj, wtan_adj, left_index=True, right_index=True)
tan_metrics(adj_data, wtan_adjusted).head()

Annualized Mean  Annualized Volatility  Annualized Sharpe Ratio
0         0.404335               0.225329                 1.794422

All aspects (mean, vol, and sharpe) are lower with this adjusted TIPS.

# 3. Allocations

<i>This section is graded for a good-faith effort by your group. Submit your write-up- along with your supporting code.

* Continue with the same data file as the previous section.

* Suppose the investor has a targeted mean excess return (per month) of $\mutarg$ = 0.01.

Build the following portfolios:

#### Equally-weighted (EW)
Rescale the entire weighting vector to have target mean $\mutarg$. Thus, the $i$ element of the weight vector is,
$$\wEW_i = \frac{1}{n}$$

#### “Risk-parity” (RP)
Risk-parity is a term used in a variety of ways, but here we have in mind setting the weight of the portfolio to be proportional to the inverse of its full-sample variance estimate. Thus, the $i$ element of the weight vector is,
$$\wRP_i = \frac{1}{\sigma_i^2}$$

#### Mean-Variance (MV)
As described in `Section 2`.


### Comparing

In order to compare all these allocation methods, rescale each weight vector, such that it has targeted mean return of $\mutarg$.

* Calculate the performance of each of these portfolios over the sample.
* Report their mean, volatility, and Sharpe ratio.
* How does performance compare across allocation methods?

#Responses - Georgia Martin

#Equally-weighted

In [13]:
target_mean = 0.01
n = len(metrics)
w_ew = pd.Series(1/n, index=metrics.index)

# Portfolio mean
mean_ew = (w_ew * metrics["mean"]).sum()

# Rescale weights so portfolio mean = target_mean
w_ew_rescaled = w_ew * (target_mean / mean_ew)

***

#Risk-parity

In [14]:
inv_var = 1 / (metrics["vol"] ** 2)
w_rp_raw = inv_var / inv_var.sum()

# Portfolio mean
mean_rp = (w_rp_raw * metrics["mean"]).sum()

# Rescale to target mean
w_rp_rescaled = w_rp_raw * (target_mean / mean_rp)

#Mean-Variance

In [15]:
w_mv = tan_weights(hw1_data)["Tangency Portfolio Weights"]

# Align weights index to stats_df to be safe
w_mv = w_mv.reindex(metrics.index)

# Compute portfolio mean
mean_mv = (w_mv * metrics["mean"]).sum()

# Rescale to target mean
w_mv_rescaled = w_mv * (target_mean / mean_mv)

#Portfolio Performance

In [16]:
def portfolio_performance(weights, metrics, returns):
    weights = weights.reindex(metrics.index)
    mu = metrics["mean"].values
    Sigma = returns.cov().values

    w = weights.values
    port_mean = w @ mu
    port_vol = np.sqrt(w.T @ Sigma @ w)
    port_sharpe = port_mean / port_vol

    return pd.Series({
        "Mean": port_mean,
        "Volatility": port_vol,
        "Sharpe": port_sharpe
    })

In [18]:
perf_ew = portfolio_performance(w_ew_rescaled, metrics, hw1_data)
perf_rp = portfolio_performance(w_rp_rescaled, metrics, hw1_data)
perf_mv = portfolio_performance(w_mv_rescaled, metrics, hw1_data)

perf_df = pd.DataFrame({
    "Equally Weighted": perf_ew,
    "Risk Parity": perf_rp,
    "Mean Variance": perf_mv
}).T

print(perf_df)

                  Mean  Volatility    Sharpe
Equally Weighted  0.01    0.006487  1.541505
Risk Parity       0.01    0.011241  0.889619
Mean Variance     0.01    0.004035  2.478073


It makes sense that the MV portfolio has the highest Sharpe since it maximizes it by construction. The RP portfolio tends to control risk contributions, so it is a bit strange that it has the highest volatility. Equally weighted remains a simpler and more robust option.



# 4. EXTRA: Out-of-Sample Performance

<i>This section is not graded, and you do not need to submit it. Still, we may discuss it in class, in which case, you would be expected to know it.

### 1. One-step Out-of-Sample (OOS) Performance
Let’s divide the sample to both compute a portfolio and then check its performance out of sample.
* Using only data through the end of `2022`, compute the weights built in Section 3.
* Rescale the weights, (using just the in-sample data,) to set each allocation to have the same mean return of $\mutarg$.
* Using those weights, calculate the portfolio’s Sharpe ratio within that sample.
* Again using those weights, (derived using data through `2022`,) calculate the portfolio’s OOS Sharpe ratio, which is based only on performance in `2023-2024`.

### 2. Rolling OOS Performance

Iterate the Out-of-Sample performance every year, not just the final year. Namely,
* Start at the end of `2015`, and calculate the weights through that time. Rescale them using the mean returns through that time.
* Apply the weights to the returns in the upcoming year, (`2016`.)
* Step forward a year in time, and recompute.
* Continue until again calculating the weights through `2023` and applying them to the returns in `2024`.

Report the mean, volatility, and Sharpe from this dynamic approach for the following portfolios:
* mean-variance (tangency)
* equally-weighted
* risk-parity
* regularized

***

# 5. EXTRA: Without a Riskless Asset

<i>This section is not graded, and you do not need to submit it. Still, we may discuss it in class, in which case, you would be expected to know it.

Re-do Section 2 above, but in the model without a risk-free rate.

That is, build the MV allocation using the two-part formula in the `Mean-Variance` section of the notes.
* This essentially substitutes the risk-free rate with the minimum-variance portfolio.
* Now, the allocation depends nonlinearly on the target mean return, $\mutarg$. (With a risk-free rate, we simply scale the weights up and down to achieve the mean return.)

You will find that, conceptually, the answers are very similar.

***

# 6. EXTRA: Bayesian Allocation

Try the following allocation among the choices in `Section 3`...


#### Regularized (REG)
Much like the Mean-Variance portfolio, set the weights proportional to
$$\wREG \sim \widehat{\Sigma}^{-1}\mux$$
but this time, use a regularized covariance matrix,
$$\widehat{\Sigma} = \frac{\Sigma + \Sigma_D}{2}$$
where $\Sigma_D$ denotes a *diagonal* matrix of the security variances, with zeros in the off-diagonals.

Thus, $\widehat{\Sigma}$ is obtained from the usual covariance matrix, $\Sigma$, but shrinking all the covariances to half their estimated values.
